perf(allocator/vec2): align min amortized cap size with `std`

Dunqing opened the pull request:
Align with https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#653-656, but unfortunately, we got a performance regression from this optimization, which was caused by `let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);`. So I commented it out and added comments to describe why. Although the performance has not changed, keep the implementation the same as the standard library is also nice to have.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9857?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9857** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9857?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9857?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9856** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9856?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9855** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9855?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9752** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9752?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9742** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9742?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9679** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9655** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9730** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9730?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9857).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (approved):
@Dunqing This is a really interesting discovery that `Vec`s having a minimum capacity of 4 is a perf regression for `oxc_transformer` etc (it'd be 4 for all AST types, as they're all between 8 and 1024 bytes).

I suspect that the reason isn't the cost of the `cmp::max` call in `grow_amortized` (like your comment says), but is to do with CPU cache usage.

As you say, many `Vec`s in AST require less than 4 elements e.g. the `Vec<FormalParameter>` in `function foo(x) {}`. `FormalParameter` is 80 bytes, so if there's spare capacity of 3 in the `Vec`, that results in 240 bytes unused in the middle of the arena. With data spread out across the arena, it'll produce more L1/L2 cache misses when traversing the AST.

That's just a theory, but I doubt one `cmp::max` call could produce a significant perf regression, especially as it's on a cold path.

overlookmotel commented:
### Merge activity

* **May 3, 10:33 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(allocator/vec2): align min amortized cap size with `std` (#9857)

Align with https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#653-656, but unfortunately, we got a performance regression from this optimization, which was caused by `let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);`. So I commented it out and added comments to describe why. Although the performance has not changed, keep the implementation the same as the standard library is also nice to have.

Dunqing commented:
> @Dunqing This is a really interesting discovery that `Vec`s having a minimum capacity of 4 is a perf regression for `oxc_transformer` etc (it'd be 4 for all AST types, as they're all between 8 and 1024 bytes).
> 
> I suspect that the reason isn't the cost of the `cmp::max` call in `grow_amortized` (like your comment says), but is to do with CPU cache usage.
> 
> As you say, many `Vec`s in AST require less than 4 elements e.g. the `Vec<FormalParameter>` in `function foo(x) {}`. `FormalParameter` is 80 bytes, so if there's spare capacity of 3 in the `Vec`, that results in 240 bytes unused in the middle of the arena. With data spread out across the arena, it'll produce more L1/L2 cache misses when traversing the AST.
> 
> That's just a theory, but I doubt one `cmp::max` call could produce a significant perf regression, especially as it's on a cold path.

I agreed your point! I've tried to pre-allocate 4 or 8 capacity for most of `Vec` usages in the parser, the benchmark result surprised me, it doesn't improve but hurts 1%-2% performance. The reason is probably like you said.
